File: docs/very-old-describers/weaver.rb
class Loom::Weaver
  include Language

  def initialize context
    @actor = context[:actor]
  end

  def resolve template
    working = true
    while working
      match = template.match /\[[^\]]+\]/
      if match.present?
        template = template.sub(match[0], find_value(match[0]))
      else
        working = false
      end
    end
    template
  end

private

  def find_value token

    if token.starts_with? "[C|"
      value = cock_weaver.find_value(token)
      return value if value.present?
    end

    if token.starts_with? "[F|"
      value = frame_weaver.find_value(token)
      return value if value.present?
    end

    case token
      when "[Actor]"      then @actor.first_name
      when "[Actor's]"    then @actor.first_name_possessive
      when "[his]"        then @actor.his
      when "[His]"        then @actor.his.titlecase
      when "[him]"        then @actor.him
      when "[Him]"        then @actor.him.titlecase
      when "[He]"         then @actor.he.titlecase
      when "[he]"         then @actor.he
      when "[elf]"        then @actor.race.name
      when "[unholy]"     then random_unholy_adjective
      when "[Male Demon]" then random_male_demon_name
      else raise "Unrecognized replacement token : #{token}"
    end
  end

  def cock_weaver
    @cock_weaver ||= Loom::CockWeaver.new(@actor)
  end

  def frame_weaver
    @frame_weaver ||= Loom::FrameWeaver.new(@actor)
  end

  def random_unholy_adjective
    Roll.random_from ["unholy","demonic","infernal","satanic","corrupt",
      "blasphemous","cursed","accursed"]
  end

  def random_male_demon_name
    Roll.random_from ["Abaddon","Baal","Baphomet","Behemoth","Lucifer",
      "Maldrasen","Mephistopheles","Satan"]
  end

end

File: engine/weavers/weaver.js
global.Weaver = (function() {

  // Shortcut function for when you want to do some token replacement on a
  // single string and you only have a single character to put in the context.
  // Don't do this in a loop, adding a character looks up all of their body
  // parts and shit after all.
  async function weaveWithCharacter(text, key, character) {
    const context = new WeaverContext();
    await context.addCharacter(key, character);

    return weave(text, context);
  }

  // The weave function takes a string and recursivly itterates over it,
  // looking for {{ templates }} to replace with actual values. The context
  // defines things like which actors are in which position and so forth. If
  // for instance you're having a conversation between three prople the player
  // may be {{P}}, while the other two actors might be {{1}} and {{2}}.
  //
  // Everything that the weaver uses should already be in the context, that way
  // this function can remain synchronous.
  function weave(text, context) {
    var working = true;

    while(working) {
      var actorMatch = text.match(/{{([^}]+)::([^}]+)}}/);
      var utilityMatch = text.match(/{{([^}]+)\|([^}]+)}}/);
      var simpleMatch = text.match(/{{([^}]+)}}/);

      if (actorMatch) {
        text = text.replace(actorMatch[0], actorValue(actorMatch[1].trim(), actorMatch[2].trim(), context));
      } else if (utilityMatch) {
        text = text.replace(utilityMatch[0], utilityValue(utilityMatch[1].trim(), utilityMatch[2].trim(), context));
      } else if (simpleMatch) {
        text = text.replace(simpleMatch[0], simpleValue(simpleMatch[1].trim()));
      } else {
        working = false;
      }
    }

    return text;
  }

  function actorValue(subject, token, context) {
    if (token.startsWith('body'))      { return Weaver.BodyLoom.findValue(subject, token, context); }
    if (token.startsWith('character')) { return Weaver.CharacterLoom.findValue(subject, token, context); }
    if (token.startsWith('cock'))      { return Weaver.CockLoom.findValue(subject, token, context); }
    if (token.startsWith('gender'))    { return Weaver.GenderLoom.findValue(subject, token, context); }
    if (token.startsWith('nipples'))   { return Weaver.NipplesLoom.findValue(subject, token, context); }
    if (token.startsWith('species'))   { return Weaver.SpeciesLoom.findValue(subject, token, context); }
    if (token.startsWith('tits'))      { return Weaver.TitsLoom.findValue(subject, token, context); }
    return error(`BadToken(${subject}::${token})`);
  }

  function utilityValue(utility, argument, context) {
    if (utility.toLowerCase().startsWith('flag')) { return Weaver.FlagLoom.findValue(utility, argument, context); }
    if (utility.toLowerCase().startsWith('random')) { return Weaver.RandomLoom.findValue(utility, argument); }
    return error(`BadToken(${utility}|${argument})`);
  }

  function simpleValue(token) {
    if (token == 'sheath')    { return Random.from(['sheath','sheath','cocksheath']); }
    if (token == 'tit')       { return Random.from(['breast','tit']); }
    if (token == 'tits')      { return Random.from(['breasts','tits']); }
    if (token == 'throb')     { return Random.from(['throb','pulse']); }
    if (token == 'throbbing') { return Random.from(['throbbing','pulsing']); }
    if (token == 'wide')      { return Random.from(['wide','thick']); }
    if (token == 'unholy')    { return Random.from(['unholy','demonic','infernal','satanic','corrupt','blasphemous','cursed','accursed']); }
    if (token == 'maleDemon') { return Random.from(['Abaddon','Baal','Baphomet','Behemoth','Lucifer','Maldrasen','Mephistopheles','Satan','Slaanesh']); }

    return error(`BadToken(${token})`);
  }

  function error(message) {
    return `<span class='weaver-error error'>Error(${message})</span>`;
  }

  return { weaveWithCharacter, weave, error };

})();
